src/names.c

```diff
@@ -708,17 +708,32 @@ SEXP vec_set_rownames(SEXP x, SEXP names) {
   return x;
 }
 
+SEXP vec_set_df_rownames(SEXP x, SEXP names) {
+  x = PROTECT(r_maybe_duplicate(x));
+
+  if (names == R_NilValue) {
+    init_compact_rownames(x, vec_size(x));
```

Comment:
It might be worth special casing the `names == R_NilValue` case. If we are setting row names to `NULL` and we already have compact row names, we don't need to duplicate the input. We just return it as is. Both the array rownames setter and the atomic name setter make this check, so I think adding it for data frames would be nice to be consistent